File: PY-SDK/public/src/pyaver/host.py
from .aver_client import AverClient
from solana.publickey import PublicKey
from .data_classes import HostState
# from constants import AVER_PROGRAM_ID_DEVNET_2
from .constants import AVER_PROGRAM_ID
from .utils import sign_and_send_transaction_instructions
from solana.system_program import SYS_PROGRAM_ID
from anchorpy import Context
from solana.keypair import Keypair
from solana.rpc.types import TxOpts


class Host():
    """
    Host Class

    Hosts spinup their own frontends and collect fees in return
    """

    pubkey: PublicKey
    """
    Host public key
    """

    host_state: HostState
    """
    Host state containing parsed data
    """

    def __init__(self, pubkey: PublicKey, host_state: HostState):
        """
        Initialises Host object. Do not use this function; use Host.load() instead

        Args:
            pubkey (PublicKey): Host public key
            host_state (HostState): Host state containing parsed data
        """
        self.pubkey = pubkey
        self.host_state = host_state
    
    @staticmethod
    async def load(aver_client: AverClient, pubkey: PublicKey):
        """
        Initialises a Host object.

        Args:
            aver_client (AverClient): AverClient object
            pubkey (PublicKey): Public key of host

        Returns:
            Host: Host
        """
        host_result = await aver_client.program.account['Host'].fetch(pubkey)
        return Host(pubkey, host_result)


    @staticmethod
    def make_create_host_account_instruction(
        aver_client: AverClient,
        owner: Keypair,
        fee_payer: Keypair,
        referrer_fee_rate_offerer_bps: int = 0,
        program_id = AVER_PROGRAM_ID
    ):
        """
        Creates instruction for host account.

        Returns TransactionInstruction object only. Does not send transaction.

        Args:
            aver_client (AverClient): AverClient object
            owner (Keypair): Keypair of owner of host account
            fee_payer (Keypair): Keypair to pay fee for transaction
            referrer_fee_rate_offerer_bps (int, optional): Fees given to referrer. Defaults to 0.
            program_id (_type_, optional): Program public key. Defaults to AVER_PROGRAM_ID.

        Returns:
            TransactionInstruction: TransactionInstruction object
        """
        host_pubkey, bump = Host.derive_pubkey_and_bump(owner.public_key, program_id)
        return aver_client.program.instruction['init_host'](
            referrer_fee_rate_offerer_bps, 
            bump,
            ctx=Context(accounts={
                "payer": fee_payer.public_key,
                "owner": owner.public_key,
                "host": host_pubkey,
                "system_program": SYS_PROGRAM_ID,
                },
                signers=[fee_payer, owner]
            )
            
            )

    @staticmethod
    async def create_host_account(
        aver_client: AverClient,
        owner: Keypair,
        fee_payer: Keypair,
        send_options: TxOpts = None,
        referrer_fee_rate_offerer_bps: int = 0,
        program_id = AVER_PROGRAM_ID,
    ):
        """
        Creates host account.

        Sends an instructions to the on-chain program to create a new Host account.

        Args:
            aver_client (AverClient): AverClient object
            owner (Keypair): Keypair of owner of host account
            fee_payer (Keypair): Keypair to pay fee for transaction and rent cost
            send_options (TxOpts, optional): Options to specify when broadcasting a transaction. Defaults to None.
            referrer_fee_rate_offerer_bps (int, optional): Fees given to referrer. Defaults to 0.
            program_id (PublicKey, optional): Program public key. Defaults to AVER_PROGRAM_ID.

        Returns:
            RPCResponse: Response
        """
        ix = Host.make_create_host_account_instruction(
            aver_client,
            owner,
            fee_payer,
            referrer_fee_rate_offerer_bps,
            program_id
        )

        return await sign_and_send_transaction_instructions(
            aver_client,
            [owner],
            fee_payer,
            [ix,],
            send_options
        )

    @staticmethod
    def parse_host_state(aver_client: AverClient, buffer: bytes):
        """
        Parses raw onchain data to HostState object

        Args:
            aver_client (AverClient): AverClient object
            buffer (bytes): Raw bytes coming from onchain

        Returns:
            HostState: HostState object
        """
        host_account_info = aver_client.program.account['Host'].coder.accounts.decode(buffer)
        return host_account_info
    
    @staticmethod
    def derive_pubkey_and_bump(owner: PublicKey, program_id: PublicKey = AVER_PROGRAM_ID):
        """
        Derives PDA for host public key

        Args:
            owner (PublicKey): Owner of host account
            program_id (PublicKey, optional): Program public key. Defaults to AVER_PROGRAM_ID.

        Returns:
            PublicKey: Host public key
        """
        return PublicKey.find_program_address([bytes('host', 'utf-8'), bytes(owner)], program_id)

File: TS-SDK/admin/other_files/host.ts
import { Idl, IdlTypeDef } from "@project-serum/anchor/dist/cjs/idl"
import {
  IdlTypes,
  TypeDef,
} from "@project-serum/anchor/dist/cjs/program/namespace/types"
import { Keypair, PublicKey, SendOptions, SystemProgram } from "@solana/web3.js"
import { AverClient } from "../../public/src/aver-client"
import { AVER_PROGRAM_ID } from "../../public/src/ids"
import { HostState } from "../../public/src/types"
import { signAndSendTransactionInstructions } from "../../public/src/utils"
import BN from "bn.js"

export class Host {
  private _pubkey: PublicKey

  private _hostState: HostState

  private constructor(pubkey: PublicKey, hostState: HostState) {
    this._pubkey = pubkey
    this._hostState = hostState
  }

  /**
   * Load the host
   *
   * @param {AverClient} averClient The aver client instance
   * @param {PublicKey} pubkey The hosts pubkey
   *
   * @returns {Host}
   */
  static async load(averClient: AverClient, pubkey: PublicKey) {
    const program = averClient.program
    const hostResult = await program.account["host"].fetch(pubkey.toBase58())
    const hostState = Host.parseHostState(hostResult)

    return new Host(pubkey, hostState)
  }

  /**
   * Function to format the instruction to create a host account
   *
   * @param {AverClient} averClient The aver client instance
   * @param {PublicKey} owner The host owner
   * @param {PublicKey} feePayer The fee payer
   * @param {BN} referrerFeeRateOfferedBps Referrer fee rate
   * @param {PublicKey} programId The program Id
   *
   * @returns {Promise<TransactionInstruction>}
   */
  static async makeCreateHostAccountInstruction(
    averClient: AverClient,
    owner?: PublicKey,
    feePayer?: PublicKey,
    referrerFeeRateOfferedBps: BN = new BN(0),
    programId = AVER_PROGRAM_ID
  ) {
    const program = averClient.program
    const hostOwner = owner || averClient.owner
    const hostFeePayer = feePayer || hostOwner
    const [hostPubkey, bump] = await Host.derivePubkeyAndBump(
      hostOwner,
      programId
    )

    return program.instruction["initHost"](referrerFeeRateOfferedBps, bump, {
      accounts: {
        payer: hostFeePayer,
        owner: hostOwner,
        host: hostPubkey,
        systemProgram: SystemProgram.programId,
      },
    })
  }

  /**
   * Signs and sends the instruction to create a host account
   *
   * @param {AverClient} averClient The aver client instance
   * @param {PublicKey} owner The host owner
   * @param {PublicKey} feePayer The fee payer
   * @param {BN} referrerFeeRateOfferedBps Referrer fee rate
   * @param {SendOptions} sendOptions Options for sending the transaction
   * @param {number} manualMaxRetry The number of times to retry the transaction before failure
   * @param {PublicKey} programId The program Id
   *
   * @returns {Promise<string>}
   */
  static async createHostAccount(
    averClient: AverClient,
    owner: Keypair = averClient.keypair,
    feePayer: Keypair,
    referrerFeeRateOfferedBps: BN = new BN(0),
    sendOptions?: SendOptions,
    manualMaxRetry?: number,
    programId = AVER_PROGRAM_ID
  ) {
    const ix = await Host.makeCreateHostAccountInstruction(
      averClient,
      owner.publicKey,
      feePayer.publicKey,
      referrerFeeRateOfferedBps,
      programId
    )

    return signAndSendTransactionInstructions(
      averClient,
      [owner, feePayer],
      feePayer,
      [ix],
      sendOptions,
      manualMaxRetry
    )
  }

  private static parseHostState(
    hostResult: TypeDef<IdlTypeDef, IdlTypes<Idl>>
  ): HostState {
    return hostResult as HostState
  }

  /**
   * Derive the host public key based on the owner and the program
   *
   * @param {PublicKey} owner The host owner
   * @param {PublicKey} programId The program Id
   *
   * @returns {PublicKey}
   */
  static async derivePubkeyAndBump(
    owner: PublicKey,
    programId = AVER_PROGRAM_ID
  ) {
    return PublicKey.findProgramAddress(
      [Buffer.from("host", "utf-8"), owner.toBuffer()],
      programId
    )
  }

  // TODO
  makeCollectRevenueShareInstruction() {}

  // TODO
  async collectRevenueShare() {}

  get pubkey() {
    return this._pubkey
  }

  get owner() {
    return this._hostState.owner
  }

  get creationDate() {
    return new Date(this._hostState.creationDate.toNumber() * 1000)
  }

  get lastWithdrawal() {
    return new Date(this._hostState.lastWithdrawal.toNumber() * 1000)
  }

  get lastBalanceUpdate() {
    return new Date(this._hostState.lastBalanceUpdate.toNumber() * 1000)
  }

  get hostRevenueShareUncollected() {
    return this._hostState.hostRevenueShareUncollected
  }

  get hostRevenueShareCollected() {
    return this._hostState.hostRevenueShareCollected
  }

  get hostFeeRateBps() {
    return this._hostState.hostFeeRateBps
  }

  get referrerFeeRateOfferedBps() {
    return this._hostState.referrerFeeRateOfferedBps
  }

  get lastReferrerTermsChange() {
    return this._hostState.lastReferrerTermsChange
  }
}
